rustls/src/crypto/mod.rs

```diff
@@ -45,19 +45,137 @@ pub use crate::rand::GetRandomFailed;
+///
+/// - **Random** - see [`crate::crypto::CryptoProvider::fill_random()`].
+/// - **Cipher suites** - see [`crate::SupportedCipherSuite`], [`crate::Tls12CipherSuite`], and
+///   [`crate::Tls13CipherSuite`].
+/// - **Key exchange groups** - see [`crate::crypto::SupportedKxGroup`].
+/// - **Signature verification algorithms** - see [`crate::WebPkiSupportedAlgorithms`].
+/// - **Authentication key loading** - see [`crate::crypto::CryptoProvider::load_private_key()`] and
+///   [`crate::sign::SigningKey`].
+///
+/// # Example code
+/// See [provider-example/] for a full client and server example that uses
+/// cryptography from the [rust-crypto] and [dalek-cryptography] projects.
+///
+/// ```shell
+/// $ cargo run --example client | head -3
+/// Current ciphersuite: TLS13_CHACHA20_POLY1305_SHA256
+/// HTTP/1.1 200 OK
+/// Content-Type: text/html; charset=utf-8
+/// Content-Length: 19899
+/// ```
+///
+/// [provider-example/]: https://github.com/rustls/rustls/tree/main/provider-example/
+/// [rust-crypto]: https://github.com/rustcrypto
+/// [dalek-cryptography]: https://github.com/dalek-cryptography
 pub trait CryptoProvider: Send + Sync + Debug + 'static {
     /// Fill the given buffer with random bytes.
+    ///
+    /// The bytes must be cryptographically strong.
+    ///
+    /// This is used for all randomness required by rustls, but not necessarily
+    /// randomness required by the underlying cryptography library.  For example:
+    /// [`crate::crypto::SupportedKxGroup::start()`] requires random material to generate
+    /// an ephemeral key exchange key, but this is not included in the interface with
+    /// rustls: it is assumed that the cryptography library provides for this itself.
     fn fill_random(&self, buf: &mut [u8]) -> Result<(), GetRandomFailed>;
 
     /// Provide a safe set of cipher suites that can be used as the defaults.
+    ///
+    /// This is used by [`crate::ConfigBuilder::with_safe_defaults()`] and
+    /// [`crate::ConfigBuilder::with_safe_default_cipher_suites()`].
+    ///
+    /// Other (non-default) cipher suites can be provided separately and configured
+    /// by passing them to [`crate::ConfigBuilder::with_cipher_suites()`]
     fn default_cipher_suites(&self) -> &'static [suites::SupportedCipherSuite];
 
     /// Return a safe set of supported key exchange groups to be used as the defaults.
+    ///
+    /// This is used by [`crate::ConfigBuilder::with_safe_defaults()`] and
+    /// [`crate::ConfigBuilder::with_safe_default_kx_groups()`].
+    ///
+    /// Other (non-default) key exchange groups can be provided separately and configured
+    /// by passing them to [`crate::ConfigBuilder::with_kx_groups()`].
     fn default_kx_groups(&self) -> &'static [&'static dyn SupportedKxGroup];
 
     /// Decode and validate a private signing key from `key_der`.
     ///
+    /// This is used by [`crate::ConfigBuilder::with_client_auth_cert()`], [`crate::ConfigBuilder::with_single_cert()`],
```

Comment:
Should we also explain in docs for `with_client_auth_cert()`, `with_single_cert()` and `with_single_cert_with_ocsp()` that the key type must be supported by the `CryptoProvider`?

Reply:
Thinking about Jsha's feedback from #1603, should we also mention what the crate has typically provided to help guide what a provider should offer in a minimal support set to avoid confusion?

> While you're working on documentation in this area, [load_private_key](https://docs.rs/rustls/0.22.0-alpha.4/rustls/crypto/trait.CryptoProvider.html#tymethod.load_private_key) should probably specify which private key formats an implementation should minimally accept.

Reply:
in those places it does say "the precise set of supported key types and parameters is defined by the selected `CryptoProvider`", but will reword this a bit